cadc-pkg-server/src/main/java/org/opencadc/pkg/server/PackageRunner.java

```diff
@@ -0,0 +1,291 @@
+ * Abstract class PackageRunner: a JobRunner implementation that provides core
+ * functionality for building CADC package files (zip or tar files, for example.)
+ * Job state management and error reporting back to the Job are provided.
+ * <p>
+ * To make a child class, implement these:
+ * - Iterator PackageItem getItems(): Will be used to download files to the package
+ * - String getPackageName(): filename used to create the package.
+ * </p>
+ */
+public abstract class PackageRunner implements JobRunner {
+    private static final Logger log = Logger.getLogger(PackageRunner.class);
+
+    protected Job job;
+
+    protected JobUpdater jobUpdater;
+    protected SyncOutput syncOutput;
+    protected ByteCountOutputStream outputStreamCounter;
+    protected WebServiceLogInfo logInfo;
+    protected TarWriter writer;
+
+    public PackageRunner() {}
+
+    /**
+     * Get filename for package to be created.
+     * @return String with package name.
+     */
+    protected abstract String getPackageName();
+
+    /**
+     * Build an Iterator of PackageItem. Use the list of files provided
+     * in the Job instance (generated in the base JobRunner class.)
+     * @return PackageItem Iterator instance - populated with references to the files in Job provided.
+     */
+    protected abstract Iterator<PackageItem> getItems();
+
+    @Override
+    public void setJob(Job job) {
+        this.job = job;
+    }
+
+    @Override
+    public void setJobUpdater(JobUpdater ju) {
+        this.jobUpdater = ju;
+    }
+
+    @Override
+    public void setSyncOutput(SyncOutput so) {
+        this.syncOutput = so;
+    }
+
+    @Override
+    public void run() {
+        logInfo = new JobLogInfo(job);
+        log.info(logInfo.start());
+        long start = System.currentTimeMillis();
+
+        doIt();
+
+        if (outputStreamCounter != null) {
+            logInfo.setBytes(outputStreamCounter.getByteCount());
+        }
+        logInfo.setElapsedTime(System.currentTimeMillis() - start);
+        log.info(logInfo.end());
+    }
+
+    private void doIt() {
+        ExecutionPhase ep;
+        writer = null;
+        try {
+            ep = jobUpdater.setPhase(job.getID(), ExecutionPhase.QUEUED, ExecutionPhase.EXECUTING, new Date());
+
+            if (!ExecutionPhase.EXECUTING.equals(ep)) {
+                ep = jobUpdater.getPhase(job.getID());
+                log.debug(job.getID() + ": QUEUED -> EXECUTING [FAILED] -- DONE");
+                logInfo.setSuccess(false);
+                logInfo.setMessage("Could not set job phase to executing, was: " + ep);
+                return;
+            }
+            log.debug(job.getID() + ": QUEUED -> EXECUTING [OK]");
+
+            // Build an iterator of PackageItem from the files named
+            // in the local Job instance
+            Iterator<PackageItem> packageItems = getItems();
+
+            // TarWriter needs an output stream instantiated.
+            initOutputStream(getPackageName());
```

Comment:
the `new TarWriter` needs to be assigned to a variable outside the try/catch in order for the finally to be able to close it.

In the original code it was not a class member variahble,  just a local var declared before the try. It's always best practice to declare as locally as possible, so up on line~163

Reply:
done